File: Graph Theory Peaks/DFS/Q3. Shortest Path with Alternating Colors/Q3_ShortestPathwithAlternatingColors.py
from enum import Enum
import collections

# Enum to represent edge colors
class Color(Enum):
    kInit = 0   # Initial state (no previous edge)
    kRed = 1    # Red edge
    kBlue = 2   # Blue edge


class Solution(object):
    def shortestAlternatingPaths(self, n, redEdges, blueEdges):
        """
        :type n: int
        :type redEdges: List[List[int]]
        :type blueEdges: List[List[int]]
        :rtype: List[int]
        """

        ans = [-1] * n  
        # Result array: shortest alternating distance from node 0
        # Initialized with -1 (means not visited yet)

        graph = [[] for _ in range(n)]  
        # Adjacency list: graph[u] stores (neighbor, edgeColor)

        q = collections.deque([(0, Color.kInit)])  
        # BFS queue storing (node, previousEdgeColor)
        # Start from node 0 with no previous color

        # Build graph with red edges
        for u, v in redEdges:
            graph[u].append((v, Color.kRed))

        # Build graph with blue edges
        for u, v in blueEdges:
            graph[u].append((v, Color.kBlue))

        step = 0  
        # BFS level = distance from node 0

        while q:  
            # Standard BFS traversal

            for _ in range(len(q)):  
                # Process all nodes at the current BFS level

                u, prevColor = q.popleft()  
                # Current node and previous edge color

                if ans[u] == -1:  
                    ans[u] = step  
                    # First time visiting node → shortest distance found

                # Explore neighbors of current node
                for i, (v, edgeColor) in enumerate(graph[u]):

                    # Skip if:
                    # 1. Edge already used (marked by -1)
                    # 2. Edge color is same as previous color (no alternation)
                    if v == -1 or edgeColor == prevColor:
                        continue

                    q.append((v, edgeColor))  
                    # Add neighbor with current edge color as previous color

                    graph[u][i] = (-1, edgeColor)  
                    # Mark this edge as used to prevent revisiting

            step += 1  
            # Increase distance for next BFS level

        return ans  
        # Return shortest alternating paths from node 0

File: Graph Theory Peaks/DFS/Q3. Shortest Path with Alternating Colors/Q3_ShortestPathwithAlternatingColors.java
class Solution {
    public int[] shortestAlternatingPaths(int n, int[][] red_edges, int[][] blue_edges) {
        
        int[][] g = new int[n][n];                 // Adjacency matrix to store graph with color info
        buildGraph(g, n, red_edges, blue_edges);   // Build graph with encoded edge colors
        
        Queue<int[]> q = new LinkedList<>();       // Queue for BFS traversal
        
        q.offer(new int[]{0, 1});                  // Start from node 0 assuming previous edge was RED
        q.offer(new int[]{0, -1});                 // Start from node 0 assuming previous edge was BLUE
        
        int len = 0;                               // Distance (number of edges from node 0)
        
        int[] res = new int[n];                    // Result array for shortest distance
        Arrays.fill(res, Integer.MAX_VALUE);       // Initialize all distances as infinity
        res[0] = 0;                                // Distance to itself is 0
        
        Set<String> visited = new HashSet<>();     // Track visited states (node + previous color)
        
        while (!q.isEmpty()) {                     // Standard BFS loop
            int size = q.size();                   // Number of nodes at current level
            len++;                                 // Increase distance level
            
            for (int i = 0; i < size; i++) {       // Process all nodes at current BFS level
                int[] cur = q.poll();              // Get current state
                int node = cur[0];                 // Current node
                int color = cur[1];                // Previous edge color
                
                int oppoColor = -color;            // Required next edge color (alternate color)
                
                for (int j = 1; j < n; j++) {      // Check all possible neighbors
                    if (g[node][j] == oppoColor || // Valid if edge has opposite color
                       g[node][j] == 0) {          // Or if both colors exist
                    
                        if (!visited.add(j + "" + oppoColor)) continue; 
                        // Skip if this (node + color state) was already visited
                        
                        q.offer(new int[]{j, oppoColor}); 
                        // Add neighbor with new edge color
                    
                        res[j] = Math.min(res[j], len); 
                        // Update shortest distance
                    }
                }
            }
        }
        
        for (int i = 1; i < n; i++) {               // Convert unreachable nodes
            if (res[i] == Integer.MAX_VALUE) {     // If still infinity
                res[i] = -1;                       // Mark as unreachable
            }
        }
        
        return res;                                // Return shortest alternating distances
    }
    
    private void buildGraph(int[][] g, int n, int[][] red_edges, int[][] blue_edges) {
        
        for (int i = 0; i < n; i++) {               // Initialize graph matrix
            Arrays.fill(g[i], -n);                  // Default value meaning no edge
        }
        
        for (int[] e : red_edges) {                 // Process all red edges
            int from = e[0];                        // Source node
            int to = e[1];                          // Destination node
            g[from][to] = 1;                        // Mark red edge as 1
        }
        
        for (int[] e : blue_edges) {                // Process all blue edges
            int from = e[0];                        // Source node
            int to = e[1];                          // Destination node
            
            if (g[from][to] == 1) {                 // If red edge already exists
                g[from][to] = 0;                    // Mark as both colors available
            } else {
                g[from][to] = -1;                   // Otherwise mark as blue edge
            }
        }
    }
}